### ISSUE TITLE
Help drivers to invalidate tablets in time

### ISSUE THREAD

**Issue by dkropachev** [scylladb/scylladb#21664]
## Abstract
Driver's tablet information is lagging behind.
In order to keep `tablet records` updated drivers imploy two methods:
1. `Tablet records` invalidation happens on `SCHEMA_CHANGE` or `TOPOLOGY_CHANGE` events or on `control connection` reconnect.
Driver go over known `tablet records` and remove records that are affected by the change.
2. When a node receives query to the host/shard that does not belong to the tablet replica set, it adds extra payload to the response including `tablet information` for token from the query, drivers pick up this payload and update their `tablet records`. 

But even with all that there are cases when replica set could be changed on cluster side, but driver will not notice it even after executing a query for target token:
### 1. Driver preserve replica set on query planing 
In some cases (Say for `LWT` queries) drivers preserve order of replica set from the tablet or vnodes, when it scheduled a query. In such case driver will get replica record information from cluster only if it runs query on the replica that is not in the replica set anymore. I.e. if replica set was [A,B,C] then got changed to [A,B,D] then driver will try to execute query on A, then on B, if any query succeed driver will not get any `tablet record` information in the response, and only if fails on both attempts it proceed with `C` and will  get `tablet record` information in the response, which is highly unlikely.

Worst-case scenario for `LWT` queries:
Cluster size = `3`, RF = `3` CL = `Quorum`
Say a tablet have following replica set: `[A, B, C]`
Tablet migration happened and now replica set is `[B, A, ...`

Drivers that manage to get new data will route most `LWT` queries to `B` node, while drivers that have old `tablet record`, will schedule nodes to node `A` and will never receive an updated `tablet record`.

As a result, It will cause excessive congestion (you can read more about it in `LWT query planning issue` references).

### 2. Replica set gets number of replicas increased
When RF is increased, it is expected that tablet migration will add one more record into `replica set`.
Since the driver will target old replicas, it will never get tablet record information from the cluster and will never recognize new replicas.

As result, new node will not get any traffic from the driver

## Solutions
There are some solutions on the driver side that can fix given use cases, all of them have downsides.
Best solution with minimal downsides will require server side changes. 

Options that were discussed before:
### 1. Include tablet version/generation/hash into request custom payload

Drivers can populate custom payload `target_tablet_generation` with target tablet id for certain frames.
It will be picked up by node and if tablet generation is different, node would add updated tablet info into response.

Drawbacks:
Custom payload is a tuple of (key_len<int>, key<text>, value_len<int>, value<text>) records.
Prepared statement execution frame for simple query `SELECT pk, ck, v FROM test1.table1 WHERE pk = ?;` where `pk, ck and v` are `int` takes `52 bytes` with customPayload being `nil`.

Adding a `target_tablet_generation` as string of two characters into such frame increases size by `2` (size of custom payload lenght) + `4` (size of key lengh) + `24` (lenght of key `target_tablet_generation`) + `4` (size of value length) + `2` (value lenght), in total of `36` bytes.
Which is `70%` increase for given query

### 2. Include tablet version/generation/hash into request frame body
Have another request flag, that is currently not occupied:
1. Cassandra: https://github.com/apache/cassandra/blob/2facddc33ebc64618a3b0557752ca638497f20b2/doc/native_protocol_v4.spec#L125-L158
2. Scylla: https://github.com/scylladb/scylladb/blob/0ea0234a7a7415731dc447523b4a324f17ce74b6/transport/server.hh#L80-L85

If the flag is set, node reads a `tablet generation/version/hash` at the end of the frame and if `generation/version/hash` is different from what node has, it adds tablet information to the response. 

Since drivers can address nodes that do not support this feature, there should be feature negotiation on control connection, in the same manner as it is currently implemented for tablets.

Overhead is only 1 byte.

### 3. Another event type
Make node to send out special events that would signal the driver to invalidate tablets for given scope: ks/table/tablet/token range ..etc.

Probably will require feature signaling from server side, to avoid double invalidation on driver side.

## References
- LWT query planning issue: [scylla-drivers#20](https://github.com/scylladb/scylla-drivers/issues/20), [java-driver#24](https://github.com/scylladb/java-driver/issues/24), [gocql#40](https://github.com/scylladb/gocql/issues/40)
- Keeping drivers `tablet records` updated: [scylla-drivers#27](https://github.com/scylladb/scylla-drivers/issues/27), [scylla-drivers#30](https://github.com/scylladb/scylla-drivers/issues/30)

**Comment by dkropachev** [comment]
There is another solution for `LWT case` of `Driver preserve replica set on query planing` proposed by @Lorak-mmk at https://github.com/scylladb/scylla-drivers/issues/27#issuecomment-2460621375, to make server include tablet information for LWT requests if request lands on any node/shard that is not primary replica.

Pros:
1. It is easier than any proposed solutions
2. Does not require any changes on driver side 

Cons:
1. Limited scope, does not fix non-LWT cases.
2. Not future proof. If we eventually come up with LWT load balancing (POC: [java-driver#198](https://github.com/scylladb/java-driver/pull/198)) it will create an excessive network traffic.

**Comment by dkropachev** [comment]
> Implement server-side solution for LWT (https://github.com/scylladb/scylladb/issues/21664#issuecomment-2494335462 ), ideally before LWT with tablets is released. Why? LWT is the one case that is important and at the same time difficult to handle driver-side reasonably well, and the proposed server-side fix should be quite easy to implement.

Why do you want to release it before `LWT+tablets` ? Why not to release it all together ?

> There of course should be a new feature (probably "tablets-routing-v2") negotiated at startup.
> If this feature is enabled:
> EXECUTE message gets additional byte at the end. This byte is basically a flag. If it is enabled then N (probably 8) bytes follow it, and they contain hash / id of the tablet that was used by the driver to make the query.
> So EXECUTE body, which previously was:
> <id><query_parameters>
> becomes
> <id><query_parameters><tablet_flag_byte><tablet_hash>?

Great idea, I put it into `4. Include tablet version/generation/hash into request frame body (protocol extention)`

> What should be the id? I'm not sure. Do the tablets have some kind of unique id? If so, it could be used. If not, then hashing first token, last token, and a list of replicas into some 64 bit hash should be good enough.

WDYT of `generation`, i.e. just a byte that get increased by 1 every time when tablet is changed ?

**Comment by Lorak-mmk** [comment]
> > Implement server-side solution for LWT ([#21664 (comment)](https://github.com/scylladb/scylladb/issues/21664#issuecomment-2494335462) ), ideally before LWT with tablets is released. Why? LWT is the one case that is important and at the same time difficult to handle driver-side reasonably well, and the proposed server-side fix should be quite easy to implement.
> 
> Why do you want to release it before `LWT+tablets` ? Why not to release it all together ?

I said to implement it before LWT+Tablets is released. The goal is to not release LWT + Tablets without drivers fix.

> WDYT of generation, i.e. just a byte that get increased by 1 every time when tablet is changed ?

I think the specifics of what consitutes tablet id / version / hash should be left to core. Intuition tells me that 1 byte "generation" will result in more edge cases and may be harder to implement (because its not possible to calculate it based on other tablet data), but I'm not sure.

**Comment by mykaul** [comment]
@bhalevy , @tgrabiec - please review and assign an owner.

**Comment by dkropachev** [comment]
> > > Implement server-side solution for LWT ([#21664 (comment)](https://github.com/scylladb/scylladb/issues/21664#issuecomment-2494335462) ), ideally before LWT with tablets is released. Why? LWT is the one case that is important and at the same time difficult to handle driver-side reasonably well, and the proposed server-side fix should be quite easy to implement.
> > 
> > 
> > Why do you want to release it before `LWT+tablets` ? Why not to release it all together ?
> 
> I said to implement it before LWT+Tablets is released. The goal is to not release LWT + Tablets without drivers fix.

Agree, @bhalevy, @tgrabiec, just to clarify on priority, from our end we don't think it needs to be fixed ASAP and does not fall into current `tablet priorityzation`, but this issue needs to be fixed alongside with LWT for tablets, including drivers counterpart.

**Comment by roydahan** [comment]
@emdroid I think your team is responsible for LWT with tablets, are you aware of this issue?

**Comment by emdroid** [comment]
@gusev-p FYI (for the LWT on tablets)

**Comment by dkropachev** [comment]
> [@gusev-p](https://github.com/gusev-p) FYI (for the LWT on tablets)

Probably it has to be added to https://github.com/scylladb/scylladb/issues/18068

**Comment by dorlaor** [comment]
Do we have a full documentation  of the CQL tablet reliance and protocol? This needs a standalone doc, not just a github

**Comment by tgrabiec** [comment]
There is documentation in [docs/dev/protocol-extensions.md](https://github.com/scylladb/scylladb/blob/master/docs/dev/protocol-extensions.md#sending-tablet-info-to-the-drivers)

**Comment by dkropachev** [comment]
@gusev-p , is this considered in https://github.com/scylladb/scylladb/issues/18068 ?

**Comment by avikivity** [comment]
This can be solved on the server side, by sending the tablet metadata even if not needed with low probability (1:10,000). Eventually every tablet on every client gets notified and starts to use the new replicas.

**Comment by dkropachev** [comment]
> This can be solved on the server side, by sending the tablet metadata even if not needed with low probability (1:10,000). Eventually every tablet on every client gets notified and starts to use the new replicas.

So you need to make 10000 requests to get one tablet updated, say you have 4k tablets, only after 80m requests at best you will have all of them updated.

**Comment by avikivity** [comment]
Is it a problem? Let's say the cluster runs 10k requests per second. Then after 8,000 seconds (2 hours) the transition will be complete.

Since adding racks/datacenters is very rare, do we care?

Note: it's actually 2 hours multiplied by the number of client processes. But again, are we in a hurry here?

**Comment by gusev-p** [comment]
> There is another solution for `LWT case` of `Driver preserve replica set on query planing` proposed by [@Lorak-mmk](https://github.com/Lorak-mmk) at [scylladb/scylla-drivers#27 (comment)](https://github.com/scylladb/scylla-drivers/issues/27#issuecomment-2460621375), to make server include tablet information for LWT requests if request lands on any node/shard that is not primary replica.
> 
> Pros:
> 
> 1. It is easier than any proposed solutions
> 2. Does not require any changes on driver side

@dkropachev What if a tablet has replicas in different DCs? We probably should have different "primary" replicas in different DCs, since `LOCAL_SERIAL` mode (which users would probably use in this case) considers only dc-local replicas.

**Comment by gusev-p** [comment]
Can't we just grab the `system.tablets` table periodically from the driver? Look at the intersection of `replicas` and `new_replicas` columns. We can also probably add a new notification `topology_change{change_type::TABLETS_CHANGED};`, trigger it in `storage_service::topology_state_load` like we do for `topology_change::new_node/ topology_change::removed_node`. Seems simpler than keep extending the CQL protocol.

**Comment by dkropachev** [comment]
> Can't we just grab the `system.tablets` table periodically from the driver? Look at the intersection of `replicas` and `new_replicas` columns. We can also probably add a new notification `topology_change{change_type::TABLETS_CHANGED};`, trigger it in `storage_service::topology_state_load` like we do for `topology_change::new_node/ topology_change::removed_node`. Seems simpler than keep extending the CQL protocol.

No we can't due to scaling concerns, if user maintains lot's of tables and/or have lot's of clients it is going to miserably fail.
Some of you users have close to 1m drivers instances targeting same cluster.

**Comment by dkropachev** [comment]
@Lorak-mmk , @gusev-p , I believe if we implement proposal from another [issue](https://github.com/scylladb/scylla-drivers/issues/38), to rotate replica set based on token it will solve problem with LWT case, since now all three replicas will be involved and if at some point query hit wrong replica driver will get tablet hint from server.

**Comment by gusev-p** [comment]
> > Can't we just grab the `system.tablets` table periodically from the driver? Look at the intersection of `replicas` and `new_replicas` columns. We can also probably add a new notification `topology_change{change_type::TABLETS_CHANGED};`, trigger it in `storage_service::topology_state_load` like we do for `topology_change::new_node/ topology_change::removed_node`. Seems simpler than keep extending the CQL protocol.
> 
> No we can't due to scaling concerns, if user maintains lot's of tables and/or have lot's of clients it is going to miserably fail. Some of you users have close to 1m drivers instances targeting same cluster.

1. We do exactly that for vnodes -- pulling entire `system.peers` from time to time - what's the difference? It might be ~10x smaller (if we have a cluster with 100 nodes vs 1000 tablets in a table), but is that important? It's still a tiny table fully served from the node's row cache.
2. If the user queries a full token range (which seems like a sensible thing to assume) the same amount of information (all tablets for a table) would need to be carried to the client if we go with your proposal, just in different queries and smaller chunks -> more overhead. Again, what's the benefit?
3. We can make drivers query `system.tablets` only for the relevant tables (for a specific `table_id` ) + subscribe to `TABLETS_CHANGED` only to those relevant tables.
4. Are there any tests or measurements that prove this "is going to miserably fail"?

**Comment by mykaul** [comment]
@gusev-p - I think there's a hidden assumption that system.peers changes are far less frequent than what might system.tablets be.
I do believe that fetching per specific table is a good initiative (and overall, I believe 'per table' for a driver is the right thing to do.

**Comment by avikivity** [comment]
system.tablets can be very large. The design point is 100k rows (and each row has collections that scale with the number of datacenters).

**Comment by gusev-p** [comment]
OK, we can:
* not trigger `topology_change{change_type::TABLETS_CHANGED, ks, cf}` for intermediate states of tablet migration, only for the final one.
* leverage the `tablet_metadata_change_hint` optimization -- this hint tells `topology_state_load` which (table_id, tokens) to reload, we can include this information into `TABLETS_CHANGED` events.

My point is, we already have events as a mechanism to communicate information from Scylla nodes to drivers, why invent another ad-hoc mechanism and extend CQL protocol for it.

**Comment by mykaul** [comment]
> system.tablets can be very large. The design point is 100k rows (and each row has collections that scale with the number of datacenters).

So perhaps not only drivers may wish to fetch per table data, but perhaps even per AZ?

**Comment by gusev-p** [comment]
> > system.tablets can be very large. The design point is 100k rows (and each row has collections that scale with the number of datacenters).
> 
> So perhaps not only drivers may wish to fetch per table data, but perhaps even per AZ?

With the optimization above a tablet migration would cause all drivers to run `select * from system.tablets where table_id = ? and last_token = ?`. Alternatively, we can just include the new replica set in `TABLETS_CHANGED` event itself, then they don't need to run any queries at all.

**Comment by avikivity** [comment]
> OK, we can:
> 
> * not trigger `topology_change{change_type::TABLETS_CHANGED, ks, cf}` for intermediate states of tablet migration, only for the final one.
> * leverage the `tablet_metadata_change_hint` optimization -- this hint tells `topology_state_load` which (table_id, tokens) to reload, we can include this information into `TABLETS_CHANGED` events.
> 
> My point is, we already have events as a mechanism to communicate information from Scylla nodes to drivers, why invent another ad-hoc mechanism and extend CQL protocol for it.

An event hits every connected client, even if they are idle.

The piggy-back tablet info response allows a driver to learn the changes just in time.

**Comment by avikivity** [comment]
> > > system.tablets can be very large. The design point is 100k rows (and each row has collections that scale with the number of datacenters).
> > 
> > 
> > So perhaps not only drivers may wish to fetch per table data, but perhaps even per AZ?
> 
> With the optimization above a tablet migration would cause all drivers to run `select * from system.tablets where table_id = ? and last_token = ?`. Alternatively, we can just include the new replica set in `TABLETS_CHANGED` event itself, then they don't need to run any queries at all.

That's not bad. But you'd still need a custom event.

**Comment by gusev-p** [comment]
Another idea: use the current `tablets-routing-v1` mechanism, but for LWT requests send the tablet info back to the driver also in case of actual LWT contention. If drivers properly handle `tablets-routing-v1` there should be exactly zero contention -> no tablet info is sent - no network overhead.
Pros:
  * Solves the LWT routing/contention problem with no CQL protocol or driver changes.
  * No excessive traffic under normal workload
  * Can be used with [LWT load balancing](https://github.com/scylladb/scylla-drivers/issues/38).
  * Comparing to the LWT-only solution from [this comment](https://github.com/scylladb/scylladb/issues/21664#issuecomment-2494335462)
    * No need to share the definition of what exact replicas are "primary/secondary/..." between scylla and drivers. This includes handling different 'primary' replicas for different DCs for `LOCAL_SERIAL`, which we would need to do in the solution above.
    * No problem if a primary replica goes down. In the solution above we would keep spamming with tablet info if the drivers redirects to a secondary replica.

Cons:
  * Doesn't cover RF changes for non-LWT workloads (nothing will force the drivers to refresh the replicas list since they don't send requests to the new replica).
  * If the driver keeps sending LWT requests to the wrong replica for some reason, we would have contention and would consume network with `tablets-routing-v1` hints for each response, which might go unnoticed by the user.

Note: The tablet replicas returned by `tablets-routing-v1` can sometimes diverge from the actual replicas used by scylla coordinators. The `tablets-routing-v1` always returns `system.tablets.replicas` column, not `system.tablets.new_replicas`. During tablet migration we switch `system.tablets.replicas` to the new replica set only at the last stage (`tablet_transition_stage::end_migration`), `tablet_transition_stage::use_new` and `tablet_transition_stage::cleanup` direct writes to the new replica set but `system.tablets.replicas` is not updated yet. This means while we're cleaning up the leaving replica we'll be returning old replicas in `tablets-routing-v1` -> the driver will keep sending requests to the leaving non-replica while cleanup is in progress. Probably not a big problem since tablet cleanup shouldn't take long.

@dkropachev @Lorak-mmk What do think?

**Comment by dkropachev** [comment]
> Another idea: use the current `tablets-routing-v1` mechanism, but for LWT requests send the tablet info back to the driver also in case of actual LWT contention. If drivers properly handle `tablets-routing-v1` there should be exactly zero contention -> no tablet info is sent - no network overhead. 

It will be hard to identify right source of contention: say to driver instances run LWT query for the same PK on different nodes, only query that got executed last got the contention, but it could be the case that first driver instance wrote to wrong node.   

> Note: The tablet replicas returned by `tablets-routing-v1` can sometimes diverge from the actual replicas used by scylla coordinators. The `tablets-routing-v1` always returns `system.tablets.replicas` column, not `system.tablets.new_replicas`. During tablet migration we switch `system.tablets.replicas` to the new replica set only at the last stage (`tablet_transition_stage::end_migration`), `tablet_transition_stage::use_new` and `tablet_transition_stage::cleanup` direct writes to the new replica set but `system.tablets.replicas` is not updated yet. This means while we're cleaning up the leaving replica we'll be returning old replicas in `tablets-routing-v1` -> the driver will keep sending requests to the leaving non-replica while cleanup is in progress. Probably not a big problem since tablet cleanup shouldn't take long.
> 
> [@dkropachev](https://github.com/dkropachev) [@Lorak-mmk](https://github.com/Lorak-mmk) What do think?

Could you please detail or refere us to the doc that describes lifecycle of the  `system.tablets.replicas` and `system.tablets.new_replicas`, is it confined within one tablet ? or there is some connection to the whole migration ?

Another queston on same topic, say that LWT is scheduled into one shard and then tablet migration happeing for the key involved, what will happen to the LWT transaction ? will migration wait LWT transaction to complete ? if not what will happen if tablet migration is completed while LWT operation is in progress ?

**Comment by gusev-p** [comment]
> It will be hard to identify right source of contention: say to driver instances run LWT query for the same PK on different nodes, only query that got executed last got the contention, but it could be the case that first driver instance wrote to wrong node.

The first driver will lose contention sooner or later and will get an updated tablet info.

**Comment by gusev-p** [comment]
> Could you please detail or refere us to the doc that describes lifecycle of the `system.tablets.replicas` and `system.tablets.new_replicas`, is it confined within one tablet ? or there is some connection to the whole migration ?

[This is the only doc](https://github.com/scylladb/scylladb/blob/master/docs/dev/topology-over-raft.md#tablet-transitions) I can think of. The logic is quite tricky, so the best doc is the code) The main function is `topology_coordinator.cc/handle_tablet_migration`.

**Comment by gusev-p** [comment]
> Another queston on same topic, say that LWT is scheduled into one shard and then tablet migration happeing for the key involved, what will happen to the LWT transaction ? will migration wait LWT transaction to complete ? if not what will happen if tablet migration is completed while LWT operation is in progress ?

LWT acquires [token_metadata_guard](https://github.com/scylladb/scylladb/blob/78ef3343333530cc9fd01ea4cc92a0c57b0b97ca/locator/tablet_metadata_guard.hh#L89) (see the long comment by the link), which blocks tablet migrations (migrations just wait). If LWT turned out to be on the wrong shard -- the CQL request is re-executed on the correct shard, see [bounce_to_shard](https://github.com/scylladb/scylladb/blob/78ef3343333530cc9fd01ea4cc92a0c57b0b97ca/cql3/statements/modification_statement.cc#L415). There could be many such 'bounces' if the tablet migrates again while the request is in flight.

**Comment by gusev-p** [comment]
> Could you please detail or refere us to the doc that describes lifecycle of the `system.tablets.replicas` and `system.tablets.new_replicas`, is it confined within one tablet ? or there is some connection to the whole migration ?

The main reason why `tablets-routing-v1` is implemented in a way that allows it to temporarily diverge from the 'real' tablet replicas is because in general there are different replica sets for reading and writing. When a tablet is migrated from replica A to replica B (suppose the replica sets change is `{A, X, Y} -> {B, X, Y}`) we first enable the so-called `write_both_read_old` mode, which implies that read replicas={A,X,Y} and write_replicas={A,B,X,Y}. This is necessary to handle streaming -- while the tablet data is streamed from A to B we want to keep receiving all writes on the new replica. Then, when the streaming is finished, we enable `write_both_read_new`, which switches the reads to {B,X,Y}, and only then the writes are turned to `{B,X,Y}` too.

For simplicity, it was decided not to bring the concept of different read and write replica sets to the drivers, so we just always return 'default' tablet replicas (`system.tablets.replicas` column). Probably it would be more appropriate to return intersection of `system.tablets.replicas` and `system.tablets.new_replicas` if `new_replicas` column is not null (i.e. the tablet is being migrated).

**Comment by dkropachev** [comment]
> For simplicity, it was decided not to bring the concept of different read and write replica sets to the drivers, so we just always return 'default' tablet replicas (`system.tablets.replicas` column). Probably it would be more appropriate to return intersection of `system.tablets.replicas` and `system.tablets.new_replicas` if `new_replicas` column is not null (i.e. the tablet is being migrated).

Returning intersection will make a problem for RF change handling ,where we will delete tablets based on replicaset lengh

**Comment by gusev-p** [comment]
> Returning intersection will make a problem for RF change handling ,where we will delete tablets based on replicaset lengh

I don't understand this. When RF changes we just run rebuild with the new set of replicas, new_replicas==replicas.

**Comment by gusev-p** [comment]
@avikivity @tgrabiec @dkropachev @Lorak-mmk I’ve [prepared a spec](https://docs.google.com/document/d/1YfNb07U9Sp0W8JLcI4CCSa8hCQRBTjdpjdUY774yjmM/edit?tab=t.0) for a new version of the tablets cache on the drivers side, please have a look. After having considered the problems mentioned in this ticked and related ones, I think we should go with the CQL extension approach proposed in this ticket, it's not really that hard to implement.

**Comment by gusev-p** [comment]
> [@avikivity](https://github.com/avikivity) [@tgrabiec](https://github.com/tgrabiec) [@dkropachev](https://github.com/dkropachev) [@Lorak-mmk](https://github.com/Lorak-mmk) I’ve [prepared a spec](https://docs.google.com/document/d/1YfNb07U9Sp0W8JLcI4CCSa8hCQRBTjdpjdUY774yjmM/edit?tab=t.0) for a new version of the tablets cache on the drivers side, please have a look. After having considered the problems mentioned in this ticked and related ones, I think we should go with the CQL extension approach proposed in this ticket, it's not really that hard to implement.

I don't think [the problems discussed in the doc](https://docs.google.com/document/d/1YfNb07U9Sp0W8JLcI4CCSa8hCQRBTjdpjdUY774yjmM/edit?tab=t.0#heading=h.a3xzm7m2dygc) are a blocker for LWT over tablets. It's not difficult to implement `tablets-routing-v2` on Scylla side, I could do that and we would release it independently.

**Comment by dkropachev** [comment]
> > [@avikivity](https://github.com/avikivity) [@tgrabiec](https://github.com/tgrabiec) [@dkropachev](https://github.com/dkropachev) [@Lorak-mmk](https://github.com/Lorak-mmk) I’ve [prepared a spec](https://docs.google.com/document/d/1YfNb07U9Sp0W8JLcI4CCSa8hCQRBTjdpjdUY774yjmM/edit?tab=t.0) for a new version of the tablets cache on the drivers side, please have a look. After having considered the problems mentioned in this ticked and related ones, I think we should go with the CQL extension approach proposed in this ticket, it's not really that hard to implement.
> 
> I don't think [the problems discussed in the doc](https://docs.google.com/document/d/1YfNb07U9Sp0W8JLcI4CCSa8hCQRBTjdpjdUY774yjmM/edit?tab=t.0#heading=h.a3xzm7m2dygc) are a blocker for LWT over tablets. It's not difficult to implement `tablets-routing-v2` on Scylla side, I could do that and we would release it independently.

Thanks @gusev-p , we had meeting on this topic, here are the bullet points:
1. As long as core maintain that case described in `Driver preserve replica set on query planing` (when first replica drift it's position in replica set from first place) is the rare (or imposible) case, we maintain that this is not a blocker for LWT functionality from our stand point
2. Since problem described in `Replica set gets number of replicas increased` is under full user or CX control, remedy of restarting application to drop tablets is considered good enough and non-blocker as long as it is properly documented and mentioned in release and/or feature (LWT+tablets) documents

Having all that we concluded that it is not required to have `tablets-routing-v2` feature for drivers to work properly with LWT+tablets, as long as:
1. Case of replica drifting is a rare or imposible case
2. It is properly documented that application/driver needs to be restarted after operation of increasing replication factor is performed.

We also decided to expedite release of the feature we call [better LWT load balancing](https://github.com/scylladb/scylla-drivers/issues/38) in Q4 as a remedy to some of the cases of `replica drifting`, in case if clients will be suffering from it.
It will be disabled by default, since it clashes with current drivers behavior, but they will be able to turn it on via configuration option.

**Comment by gusev-p** [comment]
> 1. As long as core maintain that case described in `Driver preserve replica set on query planing` (when first replica drift it's position in replica set from first place) is the rare (or imposible) case, we maintain that this is not a blocker for LWT functionality from our stand point

@avikivity [claims](https://docs.google.com/document/d/1YfNb07U9Sp0W8JLcI4CCSa8hCQRBTjdpjdUY774yjmM/edit?disco=AAABq4s6w5E) replica can't change its position in a tablet replica list with `rf-rack-valid`, but I don't fully understand why. @tgrabiec Could you maybe comment?

> Since problem described in Replica set gets number of replicas increased is under full user or CX control, remedy of restarting application to drop tablets is considered good enough and non-blocker as long as it is properly documented and mentioned in release and/or feature (LWT+tablets) documents

This problem isn’t specific to LWT over tablets, right? It also existed with LWT over vnodes.

> We also decided to expedite release of the feature we call https://github.com/scylladb/scylla-drivers/issues/38 in Q4

Great, the feature seems really useful.

Given all that, what’s your view on `tablets-routing-v2`? Do you think it’s still worth implementing independently of LWT over tablets?

**Comment by Lorak-mmk** [comment]
> This problem isn’t specific to LWT over tablets, right? It also existed with LWT over vnodes.

Not really. With vnodes we periodically fetch system.peers, so we can never get into a permanent state with wrong info about replicas for some token.

**Comment by Lorak-mmk** [comment]
> Given all that, what’s your view on tablets-routing-v2? Do you think it’s still worth implementing independently of LWT over tablets?

Imo it is. It fully solves all the issues with RF change / LWT without requiring any application restarts (so simplifies administration), while also simplifies tablet cache handling in drivers (no need to think about scenarios like replaced node).

**Comment by dkropachev** [comment]
> > Given all that, what’s your view on tablets-routing-v2? Do you think it’s still worth implementing independently of LWT over tablets?
> 
> Imo it is. It fully solves all the issues with RF change / LWT without requiring any application restarts (so simplifies administration), while also simplifies tablet cache handling in drivers (no need to think about scenarios like replaced node).

It is an ultimate (and most correct) solution to all the problems we can have with tablet invalidation, we definitely should consider implementing it in future.

**Comment by tgrabiec** [comment]
On Tue, Sep 9, 2025 at 7:33 PM Petr Gusev ***@***.***> wrote:

> *gusev-p* left a comment (scylladb/scylladb#21664)
> <https://github.com/scylladb/scylladb/issues/21664#issuecomment-3271670359>
>
>
>    1. As long as core maintain that case described in Driver preserve
>    replica set on query planing (when first replica drift it's position
>    in replica set from first place) is the rare (or imposible) case, we
>    maintain that this is not a blocker for LWT functionality from our stand
>    point
>
> @avikivity <https://github.com/avikivity> claims
> <https://docs.google.com/document/d/1YfNb07U9Sp0W8JLcI4CCSa8hCQRBTjdpjdUY774yjmM/edit?disco=AAABq4s6w5E>
> replica can't change its position in a tablet replica list with
> rf-rack-valid, but I don't fully understand why. @tgrabiec
> <https://github.com/tgrabiec> Could you maybe comment?
>

I don't think rf-rack-valid changes anything in that regard. I don't think
replicas currently change positions, since RF-- drops from the back, RF++
appends to back, and migration/rebuild replaces old replica in the ordered
vector in the same position. We plan to change RF-- to be able to drop
replica from the middle. So I wouldn't rely on it.

**Comment by dkropachev** [comment]
> On Tue, Sep 9, 2025 at 7:33 PM Petr Gusev ***@***.***> wrote:
>  *gusev-p* left a comment ([scylladb/scylladb#21664](https://github.com/scylladb/scylladb/issues/21664))
>  <[#21664 (comment)](https://github.com/scylladb/scylladb/issues/21664#issuecomment-3271670359)>
> 
> 
>     1. As long as core maintain that case described in Driver preserve
>     replica set on query planing (when first replica drift it's position
>     in replica set from first place) is the rare (or imposible) case, we
>     maintain that this is not a blocker for LWT functionality from our stand
>     point
> 
>  [@avikivity](https://github.com/avikivity) <https://github.com/avikivity> claims
>  <https://docs.google.com/document/d/1YfNb07U9Sp0W8JLcI4CCSa8hCQRBTjdpjdUY774yjmM/edit?disco=AAABq4s6w5E>
>  replica can't change its position in a tablet replica list with
>  rf-rack-valid, but I don't fully understand why. [@tgrabiec](https://github.com/tgrabiec)
>  <https://github.com/tgrabiec> Could you maybe comment?
> I don't think rf-rack-valid changes anything in that regard. I don't think
> replicas currently change positions, since RF-- drops from the back, RF++
> appends to back, and migration/rebuild replaces old replica in the ordered
> vector in the same position. We plan to change RF-- to be able to drop
> replica from the middle. So I wouldn't rely on it.
> 
> 
> Message ID: ***@***.***>
> […](#)

RF-- is not concern at the moment, so I assume we good here.

From drivers side it is very crucual that replica doesn't change it's position, so I wonder if we can enstil it in unit test, documentation, functional requirement ?
To make sure that it is not broken until `tablets-routing-v2` is implemented.

**Comment by gusev-p** [comment]
> On Tue, Sep 9, 2025 at 7:33 PM Petr Gusev ***@***.***> wrote:
>  *gusev-p* left a comment ([scylladb/scylladb#21664](https://github.com/scylladb/scylladb/issues/21664))
>  <[#21664 (comment)](https://github.com/scylladb/scylladb/issues/21664#issuecomment-3271670359)>
> 
> 
>     1. As long as core maintain that case described in Driver preserve
>     replica set on query planing (when first replica drift it's position
>     in replica set from first place) is the rare (or imposible) case, we
>     maintain that this is not a blocker for LWT functionality from our stand
>     point
> 
>  [@avikivity](https://github.com/avikivity) <https://github.com/avikivity> claims
>  <https://docs.google.com/document/d/1YfNb07U9Sp0W8JLcI4CCSa8hCQRBTjdpjdUY774yjmM/edit?disco=AAABq4s6w5E>
>  replica can't change its position in a tablet replica list with
>  rf-rack-valid, but I don't fully understand why. [@tgrabiec](https://github.com/tgrabiec)
>  <https://github.com/tgrabiec> Could you maybe comment?
> I don't think rf-rack-valid changes anything in that regard. I don't think
> replicas currently change positions, since RF-- drops from the back, RF++
> appends to back, and migration/rebuild replaces old replica in the ordered
> vector in the same position. We plan to change RF-- to be able to drop
> replica from the middle. So I wouldn't rely on it.
> 
> 
> Message ID: ***@***.***>
> […](#)

In my doc, I describe a specific case where this could happen: if the load balancer moves a tablet off a node and then later moves it back, while some driver still holds an outdated replica list in its tablets cache. In that case, the drivers would see the same replicas but in a different order, causing LWTs to be coordinated by different nodes → contention. I don’t think this is a major problem (it should be extremely rare, if it can happen at all), but @tgrabiec is there anything that would prevent the load balancer from moving tablets in this way?

**Comment by gusev-p** [comment]
> Not really. With vnodes we periodically fetch system.peers, so we can never get into a permanent state with wrong info about replicas for some token.

How does this work? Suppose we have a tablet with replicas {A, B} and we increase the RF from 2 to 3. Something should force the table cache entry to be invalidated, but replicas A and B are still valid. Do we just drop the cache periodically?

**Comment by Lorak-mmk** [comment]
> How does this work? Suppose we have a tablet with replicas {A, B} and we increase the RF from 2 to 3. Something should force the table cache entry to be invalidated, but replicas A and B are still valid. Do we just drop the cache periodically?

I'm confused. I was talking about vnodes, in response to your message about LWT over vnodes, but now you are asking about tablets.
With vnodes drivers don't have "cache". They fetch system.peers (periodically) which contains all informations needed to calculate replicas for given token.
With tablets, drivers rely on tablet feedback (tablets-routing-v1), which can become stale, with no good way for the driver to detect that.

**Comment by gusev-p** [comment]
> Since problem described in Replica set gets number of replicas increased is under full user or CX control, remedy of restarting application to drop tablets is considered good enough and non-blocker as long as it is properly documented and mentioned in release and/or feature (LWT+tablets) documents

@dkropachev This looks like more of an issue for regular eventually consistent workloads over tablets (since they wouldn’t see the new replicas), not for LWT, since @tgrabiec mentioned above, RF++ just appends to the end of the replica list. Do we mention this in the tablets user docs?

**Comment by dkropachev** [comment]
> > Since problem described in Replica set gets number of replicas increased is under full user or CX control, remedy of restarting application to drop tablets is considered good enough and non-blocker as long as it is properly documented and mentioned in release and/or feature (LWT+tablets) documents
> 
> @dkropachev This looks like more of an issue for regular eventually consistent workloads over tablets (since they wouldn’t see the new replicas), not for LWT, since @tgrabiec mentioned above, RF++ just appends to the end of the replica list. Do we mention this in the tablets user docs?

Correct, it goes for both LWT and regular queries.

**Comment by gusev-p** [comment]
> > > Since problem described in Replica set gets number of replicas increased is under full user or CX control, remedy of restarting application to drop tablets is considered good enough and non-blocker as long as it is properly documented and mentioned in release and/or feature (LWT+tablets) documents
> > 
> > 
> > [@dkropachev](https://github.com/dkropachev) This looks like more of an issue for regular eventually consistent workloads over tablets (since they wouldn’t see the new replicas), not for LWT, since [@tgrabiec](https://github.com/tgrabiec) mentioned above, RF++ just appends to the end of the replica list. Do we mention this in the tablets user docs?
> 
> Correct, it goes for both LWT and regular queries.

I mean it's not specific for LWT over tablets, we should mention this some general "tablets limitations" section of the user docs.

@annastuchlik When the client changes replication factor via `ALTER KEYSPACE` command, they should restart their client applications -- the drivers currently don't handle this automatically. Would it be appropriate to add a section here https://docs.scylladb.com/manual/stable/architecture/tablets.html#limitations-and-unsupported-features?

**Comment by Lorak-mmk** [comment]
@gusev-p It is described in https://docs.scylladb.com/manual/stable/kb/rf-increase.html#increasing-the-rf-in-tablets-based-keyspaces and also in https://docs.scylladb.com/manual/stable/cql/ddl.html#alter-keyspace-with-tablets

**Comment by gusev-p** [comment]
> [@gusev-p](https://github.com/gusev-p) It is described in https://docs.scylladb.com/manual/stable/kb/rf-increase.html#increasing-the-rf-in-tablets-based-keyspaces and also in https://docs.scylladb.com/manual/stable/cql/ddl.html#alter-keyspace-with-tablets

OK, so it seems we don't need any other user doc changes with regard to LWT over tablets.

**Comment by gusev-p** [comment]
I created a `tablets-routing-v2` ticket. Should we close this one?

**Comment by gusev-p** [comment]
> In my doc, I describe a specific case where this could happen: if the load balancer moves a tablet off a node and then later moves it back, while some driver still holds an outdated replica list in its tablets cache. In that case, the drivers would see the same replicas but in a different order, causing LWTs to be coordinated by different nodes → contention. I don’t think this is a major problem (it should be extremely rare, if it can happen at all), but [@tgrabiec](https://github.com/tgrabiec) is there anything that would prevent the load balancer from moving tablets in this way?

@tgrabiec this question is still relevant

**Comment by tgrabiec** [comment]
> > On Tue, Sep 9, 2025 at 7:33 PM Petr Gusev _**@**_.***> wrote:
> > _gusev-p_ left a comment ([scylladb/scylladb#21664](https://github.com/scylladb/scylladb/issues/21664))
> > <[#21664 (comment)](https://github.com/scylladb/scylladb/issues/21664#issuecomment-3271670359)>
> > ```
> > 1. As long as core maintain that case described in Driver preserve
> > replica set on query planing (when first replica drift it's position
> > in replica set from first place) is the rare (or imposible) case, we
> > maintain that this is not a blocker for LWT functionality from our stand
> > point
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > [@avikivity](https://github.com/avikivity) https://github.com/avikivity claims
> > https://docs.google.com/document/d/1YfNb07U9Sp0W8JLcI4CCSa8hCQRBTjdpjdUY774yjmM/edit?disco=AAABq4s6w5E
> > replica can't change its position in a tablet replica list with
> > rf-rack-valid, but I don't fully understand why. [@tgrabiec](https://github.com/tgrabiec)
> > https://github.com/tgrabiec Could you maybe comment?
> > I don't think rf-rack-valid changes anything in that regard. I don't think
> > replicas currently change positions, since RF-- drops from the back, RF++
> > appends to back, and migration/rebuild replaces old replica in the ordered
> > vector in the same position. We plan to change RF-- to be able to drop
> > replica from the middle. So I wouldn't rely on it.
> > Message ID: _**@**_.***>
> > […](#)
> 
> In my doc, I describe a specific case where this could happen: if the load balancer moves a tablet off a node and then later moves it back, while some driver still holds an outdated replica list in its tablets cache. In that case, the drivers would see the same replicas but in a different order, causing LWTs to be coordinated by different nodes → contention. I don’t think this is a major problem (it should be extremely rare, if it can happen at all), but [@tgrabiec](https://github.com/tgrabiec) is there anything that would prevent the load balancer from moving tablets in this way?

Depends what you mean by a replica. Do you mean host:shard pair (address), or do you mean actual replica, with a particular history of writes, which can moves across nodes. I was talking about the latter.  

Do you mean this (B changes position from 1 to 0, B->D, A->B):
{A, B, C} -> {A, D, C} -> {B, D, C}

Above, replica B (address) moves from index 1 to 0. But it's not the actual tablet replica which moves, just the address. The original replica which was on B stays at index 1, and is now on D.

**Comment by tgrabiec** [comment]
If the concern is about request being sent to the same node for a given tablet for coordination, then it's not possible to avoid temporary discrepancies during topology changes. e.g.

{A, B, C} -> {D, B, C} 

If drivers decided to consistently choose A, they have to change the decision post-migration, and now choose, for example, D.

**Comment by tgrabiec** [comment]
> > Since problem described in Replica set gets number of replicas increased is under full user or CX control, remedy of restarting application to drop tablets is considered good enough and non-blocker as long as it is properly documented and mentioned in release and/or feature (LWT+tablets) documents
> 
> [@dkropachev](https://github.com/dkropachev) This looks like more of an issue for regular eventually consistent workloads over tablets (since they wouldn’t see the new replicas), not for LWT, since [@tgrabiec](https://github.com/tgrabiec) mentioned above, RF++ just appends to the end of the replica list. Do we mention this in the tablets user docs?

No, it's an implementation detail, not a guarantee.

**Comment by gusev-p** [comment]
> If the concern is about request being sent to the same node for a given tablet for coordination, then it's not possible to avoid temporary discrepancies during topology changes. e.g.
> 
> {A, B, C} -> {D, B, C}
> 
> If drivers decided to consistently choose A, they have to change the decision post-migration, and now choose, for example, D.

I was talking about replica addresses, not the actual 'data replicas'. The concern is that the current protocol to notify drivers about tablet changes (`tablets-routing-v1`) doesn't prevent this scenario:
  1. tablet replicas is {A, B, C}, all drivers consistently send LWT requests to A (the primary replica) -> no paxos contention
  2. the replica from A migrates to another node (D) and some other replica (a replica at any position in the tablet replicas list other than zero) is migrated to A, so that we end up with e.g. this {D, A, C}
  3. Suppose the step 2 happened while some driver still holds the old tablet replicas {A, B, C}. Now when this driver makes an LWT, it goes to A, and nothing will force the node A to send `tablets-routing-v1` to the driver because it's still mentioned in the tablet replicas list.
  4. There could be many drivers like the one from 3 => some subset of drivers send LWTs to A, another subset send them to D => lots of paxos contention.

**Comment by gusev-p** [comment]
We’re trying to assess how likely this scenario is, to decide whether it’s a blocker for the LWT over tablets release. @tgrabiec, do I understand correctly that this scenario is quite likely?

**Comment by tgrabiec** [comment]
> > If the concern is about request being sent to the same node for a given tablet for coordination, then it's not possible to avoid temporary discrepancies during topology changes. e.g.
> > {A, B, C} -> {D, B, C}
> > If drivers decided to consistently choose A, they have to change the decision post-migration, and now choose, for example, D.
> 
> I was talking about replica addresses, not the actual 'data replicas'. The concern is that the current protocol to notify drivers about tablet changes (`tablets-routing-v1`) doesn't prevent this scenario:
> 
> 1. tablet replicas is {A, B, C}, all drivers consistently send LWT requests to A (the primary replica) -> no paxos contention
> 2. the replica from A migrates to another node (D) and some other replica (a replica at any position in the tablet replicas list other than zero) is migrated to A, so that we end up with e.g. this {D, A, C}
> 3. Suppose the step 2 happened while some driver still holds the old tablet replicas {A, B, C}. Now when this driver makes an LWT, it goes to A, and nothing will force the node A to send `tablets-routing-v1` to the driver because it's still mentioned in the tablet replicas list.
> 4. There could be many drivers like the one from 3 => some subset of drivers send LWTs to A, another subset send them to D => lots of paxos contention.
 
I see. With rf-rack-valid it's not possible to arrive by migration, because index 0 will switch to replicas within the same rack. It is still possible to arrive at if you change RF by removing that rack, and then change RF by adding the same rack back, which will put it at the end of the list.

Btw, if we really want to avoid contention, why not change the LWT server:
 1) pick the coordinator from replica list 
 2) if me != coordinator and !coordinator.is_down:
         send the hint back to the driver (tablets-routing-v1)
         forward the request to coordinator

That should also reduce the window of temporary discrepancies, and works with tablets-routing-v1. And protects us from bad drivers.

**Comment by gusev-p** [comment]
> With rf-rack-valid it's not possible to arrive by migration, because index 0 will switch to replicas within the same rack.

I see. The [rf-rack-valid definition](https://docs.scylladb.com/manual/stable/reference/glossary.html#term-RF-rack-valid-keyspace) implications:
  * no rack can have more than one replica
  * each slot in a tablet replica list is reserved for a particular rack
  * tablet balancer is prohibited from moving replicas between racks

This means a node `host_id` can't change its position in a tablet replica list -> no problem with LWT contention.

> It is still possible to arrive at if you change RF by removing that rack, and then change RF by adding the same rack back, which will put it at the end of the list.

This is covered by the RF change case described above -- the users are advised by the documentation to restart their client apps/drivers when RF changes -> no problem with the driver cache.

> Btw, if we really want to avoid contention, why not change the LWT server:
> 
> pick the coordinator from replica list
> if me != coordinator and !coordinator.is_down:
> send the hint back to the driver (tablets-routing-v1)
> forward the request to coordinator

In setups with multiple DCs and LOCAL_SERIAL LWTs (which, if I understand correctly, is a fairly common configuration, right @dkropachev ?), it would be more efficient to designate a separate coordinator replica in each DC.

For this to work, a Scylla node would need to determine which DC the `client_state.get_remote_address()` belongs to—but I’m not sure this information is available.

Another issue is that Scylla and the drivers must stay in strict sync on how the coordinator node is chosen. If their views diverge, Scylla might respond with `tablets-routing-v1`, while the driver continues sending the LWT to a different node. This would waste network traffic due to unnecessary `tablets-routing-v1`. A similar problem occurs if Scylla and the drivers disagree on which nodes are unavailable.

These issues make the protocol fragile. By contrast, `tablets-routing-v2` shifts responsibility to the drivers, making them the source of truth for selecting the LWT coordinator node in a most efficient way.

**Comment by dkropachev** [comment]
> We’re trying to assess how likely this scenario is, to decide whether it’s a blocker for the LWT over tablets release. [@tgrabiec](https://github.com/tgrabiec), do I understand correctly that this scenario is quite likely?

@tgrabiec , if that is true we have to make `tablets-routing-v2` into the same release.

**Comment by gusev-p** [comment]
> > We’re trying to assess how likely this scenario is, to decide whether it’s a blocker for the LWT over tablets release. [@tgrabiec](https://github.com/tgrabiec), do I understand correctly that this scenario is quite likely?
> 
> [@tgrabiec](https://github.com/tgrabiec) , if that is true we have to make `tablets-routing-v2` into the same release.

From what I understood:
* there is nothing that would stop this from happening without `rf-rack-valid`
* the problem is impossible with `rf-rack-valid`

@dkropachev So the question is, do we care enough for non `rf-rack-valid` cases?

**Comment by dkropachev** [comment]
> > > We’re trying to assess how likely this scenario is, to decide whether it’s a blocker for the LWT over tablets release. [@tgrabiec](https://github.com/tgrabiec), do I understand correctly that this scenario is quite likely?
> > 
> > [@tgrabiec](https://github.com/tgrabiec) , if that is true we have to make `tablets-routing-v2` into the same release.
> 
> From what I understood:
> * there is nothing that would stop this from happening without `rf-rack-valid`
> * the problem is impossible with `rf-rack-valid`
> 
> @dkropachev So the question is, do we care enough for non `rf-rack-valid` cases?

It depends on how and when `rf-rack-valid` going to get in.
If it is get into same release with `lwt+tablets` and enabled by default then it is ok to just document it, otherwise we have to have `tablets-routing-v2`.
Assumption here that in LWT give it more chance to play out.

On other hand, to just drop all these consideration we can go with `tablets-routing-v2` which will cover all the cases.

**Comment by gusev-p** [comment]
> It depends on how and when `rf-rack-valid` going to get in.
> If it is get into same release with `lwt+tablets` and enabled by default then it is ok to just document it, otherwise we have to have `tablets-routing-v2`.
> Assumption here that in LWT give it more chance to play out.

@mykaul Could you provide a broader perspective here? On one hand, we don’t want to block LWT over tablets on `tablets-routing-v2` (since driver-side support might take a while). On the other hand, we’re not sure about the adoption of `rf-rack-valid`. Would it be acceptable to simply state in the docs that LWT over tablets requires `rf-rack-valid` keyspaces?

**Comment by tgrabiec** [comment]
> > With rf-rack-valid it's not possible to arrive by migration, because index 0 will switch to replicas within the same rack.
> 
> I see. The [rf-rack-valid definition](https://docs.scylladb.com/manual/stable/reference/glossary.html#term-RF-rack-valid-keyspace) implications:
> 
> * no rack can have more than one replica
> * each slot in a tablet replica list is reserved for a particular rack
> * tablet balancer is prohibited from moving replicas between racks
> 
> This means a node `host_id` can't change its position in a tablet replica list -> no problem with LWT contention.
> 
> > It is still possible to arrive at if you change RF by removing that rack, and then change RF by adding the same rack back, which will put it at the end of the list.
> 
> This is covered by the RF change case described above -- the users are advised by the documentation to restart their client apps/drivers when RF changes -> no problem with the driver cache.
> 
> > Btw, if we really want to avoid contention, why not change the LWT server:
> > pick the coordinator from replica list
> > if me != coordinator and !coordinator.is_down:
> > send the hint back to the driver (tablets-routing-v1)
> > forward the request to coordinator
> 
> In setups with multiple DCs and LOCAL_SERIAL LWTs (which, if I understand correctly, is a fairly common configuration, right [@dkropachev](https://github.com/dkropachev) ?), it would be more efficient to designate a coordinator replica in each DC.
> 
> For this to work, a Scylla node would need to determine which DC the `client_state.get_remote_address()` belongs to—but I’m not sure this information is available.

You can assume that the node handling the request is in the intended DC, no? That's how LOCAL_SERIAL currently works.

> 
> Another issue is that Scylla and the drivers must stay in strict sync on how the coordinator node is chosen. If their views diverge, Scylla might respond with `tablets-routing-v1`, while the driver continues sending the LWT to a different node. This would waste network traffic due to unnecessary `tablets-routing-v1`. A similar problem occurs if Scylla and the drivers disagree on which nodes are unavailable.

If the views diverge, it means we are prone to contention. Contention is worse than wasting network, because it can easily lead to unavailability. So sacrificing some bandwidth is a good tradeoff.

Drivers should generally follow server's view of UP status, but they also make their own decisions based on errors, which can be a source of discrepancy between different clients connecting to the same cluster. Would be nice to be resilient against that.

> 
> These issues make the protocol fragile. By contrast, `tablets-routing-v2` shifts responsibility to the drivers, making them the source of truth for selecting the LWT coordinator node in a most efficient way.

I think `tablets-routing-v2` is good regardless. But even with that, you can have discrepancies due to other factors.

We have 1 scylla implementation and N driver implementations. So making drivers be the source of truth is more fragile. It would be better if the server told the drivers which coordinator to use. Like we plan to do with strongly consistent tables on raft. So we could build that into `tablets-routing-v2` already in preparation for that.

**Comment by tgrabiec** [comment]
> > > > We’re trying to assess how likely this scenario is, to decide whether it’s a blocker for the LWT over tablets release. [@tgrabiec](https://github.com/tgrabiec), do I understand correctly that this scenario is quite likely?
> > > 
> > > 
> > > [@tgrabiec](https://github.com/tgrabiec) , if that is true we have to make `tablets-routing-v2` into the same release.
> > 
> > 
> > From what I understood:
> > 
> > * there is nothing that would stop this from happening without `rf-rack-valid`
> > * the problem is impossible with `rf-rack-valid`
> > 
> > [@dkropachev](https://github.com/dkropachev) So the question is, do we care enough for non `rf-rack-valid` cases?
> 
> It depends on how and when `rf-rack-valid` going to get in.

It looks like it was already released in 2025.2 \cc @dawmd  

We can say that LWT is supported only with rf_rack_valid_keyspaces=true. We already plan to do that for materialized views on tablets.

**Comment by dkropachev** [comment]
> > With rf-rack-valid it's not possible to arrive by migration, because index 0 will switch to replicas within the same rack.
> 
> I see. The [rf-rack-valid definition](https://docs.scylladb.com/manual/stable/reference/glossary.html#term-RF-rack-valid-keyspace) implications:
> 
> * no rack can have more than one replica
> * each slot in a tablet replica list is reserved for a particular rack
> * tablet balancer is prohibited from moving replicas between racks
> 
> This means a node `host_id` can't change its position in a tablet replica list -> no problem with LWT contention.
> 
> > It is still possible to arrive at if you change RF by removing that rack, and then change RF by adding the same rack back, which will put it at the end of the list.
> 
> This is covered by the RF change case described above -- the users are advised by the documentation to restart their client apps/drivers when RF changes -> no problem with the driver cache.
> 
> > Btw, if we really want to avoid contention, why not change the LWT server:
> > pick the coordinator from replica list
> > if me != coordinator and !coordinator.is_down:
> > send the hint back to the driver (tablets-routing-v1)
> > forward the request to coordinator
> 
> In setups with multiple DCs and LOCAL_SERIAL LWTs (which, if I understand correctly, is a fairly common configuration, right [@dkropachev](https://github.com/dkropachev) ?), it would be more efficient to designate a separate coordinator replica in each DC.
> 
> For this to work, a Scylla node would need to determine which DC the `client_state.get_remote_address()` belongs to—but I’m not sure this information is available.

Let's have a meeting on this topic, we already discussed it in the mail thread, but it looks like there is still gaps we need to fill in.

**Comment by gusev-p** [comment]
> You can assume that the node handling the request is in the intended DC, no? That's how LOCAL_SERIAL currently works.

We can change the current `tablets-routing-v1` implementation in Scylla so that LWT requests also return the custom payload based on the following logic:
```
replicas = <system.tablets.replicas>;
if (request.serial_consistency == LOCAL_SERIAL) {
  // filter replicas, keeping only those in the current node’s DC, preserving order
}
if (current_node != replicas[0]) {
  // return the tablets-routing-v1 custom paylod
}
```
This behavior was not part of the original `tablets-routing-v1` protocol contract between Scylla and the drivers. As a result, some drivers (or certain versions of them) may continue sending LWT requests to the “wrong” replica. This wouldn’t reduce Paxos contention (at the end of the day, it is the driver who decides which replica to send a request to), but now `tablets-routing-v1` would be included with every LWT response, adding unnecessary network overhead.

I think it's better to negotiate a new behavior with the driver at startup, when a new control connection is established. Introducing a new `tablets-routing-v2` allows just that.

> If the views diverge, it means we are prone to contention. Contention is worse than wasting network, because it can easily lead to unavailability. So sacrificing some bandwidth is a good tradeoff.

As noted above, the mere presence of `tablets-routing-v1` in a CQL response doesn't force the driver to comply -- it can still send requests to the wrong replicas.

> We have 1 scylla implementation and N driver implementations. So making drivers be the source of truth is more fragile. It would be better if the server told the drivers which coordinator to use. Like we plan to do with strongly consistent tables on raft. So we could build that into tablets-routing-v2 already in preparation for that.

The `tablets-routing-v2` would ensure that the drivers see the most up-to-date tablets topology. Part of the `tablets-routing-v2` protocol is that for LWTs, the drivers would have to traverse the replicas in the exact provided order -- they must try the next replica iff all the previous replicas are considered down (and return to the previous replica as soon as it becomes available). Same rule could be applied for strongly consistent queries in the future.

> Let's have a meeting on this topic, we already discussed it in the mail thread, but it looks like there is still gaps we need to fill in.

I'm not sure that a meeting would help (I think we're making a great progress in async discussions in the comments), but let's have one if you think it's worth it.

**Comment by tgrabiec** [comment]
> > You can assume that the node handling the request is in the intended DC, no? That's how LOCAL_SERIAL currently works.
> 
> We can change the current `tablets-routing-v1` implementation in Scylla so that LWT requests also return the custom payload based on the following logic:
> 
> ```
> replicas = <system.tablets.replicas>;
> if (request.serial_consistency == LOCAL_SERIAL) {
>   // filter replicas, keeping only those in the current node’s DC, preserving order
> }
> if (current_node != replicas[0]) {
>   // return the tablets-routing-v1 custom paylod
> }
> ```
> 
> This behavior was not part of the original `tablets-routing-v1` protocol contract between Scylla and the drivers. As a result, some drivers (or certain versions of them) may continue sending LWT requests to the “wrong” replica. This wouldn’t reduce Paxos contention (at the end of the day, it is the driver who decides which replica to send a request to), but now `tablets-routing-v1` would be included with every LWT response, adding unnecessary network overhead.
> 
> I think it's better to negotiate a new behavior with the driver at startup, when a new control connection is established. Introducing a new `tablets-routing-v2` allows just that.

If there is a behavior change, then we indeed need v2. But is there?

> 
> > If the views diverge, it means we are prone to contention. Contention is worse than wasting network, because it can easily lead to unavailability. So sacrificing some bandwidth is a good tradeoff.
> 
> As noted above, the mere presence of `tablets-routing-v1` in a CQL response doesn't force the driver to comply -- it can still send requests to the wrong replicas.

In general, yes. But it does solve the case you brought up, where discrepancy is caused by stale replica list. 

I also proposed to re-route the request on scylla side, which should reduce contention even if the driver doesn't comply, or different driver instances make different choices.

**Comment by gusev-p** [comment]
> If there is a behavior change, then we indeed need v2. But is there?

I just described the change in the previous comment. Drivers are not obliged to send LWTs to the first replica in a tablet replica list with `tablets-routing-v1`, they could send it so some other replicas, e.g. based on [LWT better load balancing](https://github.com/scylladb/scylla-drivers/issues/38). => normal LWT traffic will produce `tablets-routing-v1` custom payload on each response, consuming network traffic.

As discussed above, we can require `rf-rack-validity` for both LWT over tablets and strongly consistent requests, which means regular tablet migrations are not a problem for LWT performance. Based on discussions in  [the doc](https://docs.google.com/document/d/1YfNb07U9Sp0W8JLcI4CCSa8hCQRBTjdpjdUY774yjmM/edit?tab=t.0#heading=h.a3xzm7m2dygc), there is still motivation for `tablets-routing-v2`:
* `Tablet migration cleanup` (such cleanups expected to be short in most cases)
* `Replica count increase`
* [Node replace](https://docs.google.com/document/d/1YfNb07U9Sp0W8JLcI4CCSa8hCQRBTjdpjdUY774yjmM/edit?disco=AAABq4s6w48) while control connection is reconnecting, probably other cases when driver could miss some Scylla events.
* Simplify driver's code -- e.g. no need to handle Scylla events at all.

These issues are not a blocker for LWT over tablets release, drivers somehow deal with them.

We can use exactly the same protocol for the future strongly consistent requests. We will _have to_ implement Scylla side request forwarding for them to work, so we will be able to reuse it for LWTs as well. (/cc @gleb-cloudius).

More specifically, three cases for the `tablets-routing-v2`:
1. **Eventually-consistent requests**. No notion of the primary replica -- no Scylla-side request forwarding. Drivers send request round robin to the nearest alive replicas. Scylla just checks the replicas hash - respond with the new replica list if changed.
2. **LWT requests**. The primary replica is chosen by the driver as in [LWT better load balancing](https://github.com/scylladb/scylla-drivers/issues/38). It seems we don't need Scylla-side request forwarding in this case.
3. **strongly consistent requests**. The primary replica is chosen by Scylla, put in the first position in the replica list we send in `tablets-routing-v2`, after we compute the hash. Do the Scylla-side request forwarding if the actual replica doesn't match.

**Comment by tgrabiec** [comment]
> > If there is a behavior change, then we indeed need v2. But is there?
> 
> I just described the change in the previous comment. Drivers are not obliged to send LWTs to the first replica in a tablet replica list with `tablets-routing-v1`, they could send it so some other replicas, e.g. based on [LWT better load balancing](https://github.com/scylladb/scylla-drivers/issues/38). => normal LWT traffic will produce `tablets-routing-v1` custom payload on each response, consuming network traffic.

I see. Then we should negotiate indeed. 

If you ask me, I would still add forwarding to avoid contention on misbehaving or old drivers. 

> 
> As discussed above, we can require `rf-rack-validity` for both LWT over tablets and strongly consistent requests, which means regular tablet migrations are not a problem for LWT performance. Based on discussions in [the doc](https://docs.google.com/document/d/1YfNb07U9Sp0W8JLcI4CCSa8hCQRBTjdpjdUY774yjmM/edit?tab=t.0#heading=h.a3xzm7m2dygc), there is still motivation for `tablets-routing-v2`:
> 
> * `Tablet migration cleanup` (such cleanups expected to be short in most cases)
> * `Replica count increase`
> * [Node replace](https://docs.google.com/document/d/1YfNb07U9Sp0W8JLcI4CCSa8hCQRBTjdpjdUY774yjmM/edit?disco=AAABq4s6w48) while control connection is reconnecting, probably other cases when driver could miss some Scylla events.
> * Simplify driver's code -- e.g. no need to handle Scylla events at all.
> 
> These issues are not a blocker for LWT over tablets release, drivers somehow deal with them.
> 
> We can use exactly the same protocol for the future strongly consistent requests. We will _have to_ implement Scylla side request forwarding for them to work, so we will be able to reuse it for LWTs as well. (/cc [@gleb-cloudius](https://github.com/gleb-cloudius)).

If you include the coordinator host id hint in the response, then yes. For strongly consistent tables it cannot be derived from replica list, it's chosen by election. If we make LWT use the same protocol, we can avoid future changes. 

> 
> More specifically, three cases for the `tablets-routing-v2`:
> 
> 1. **Eventually-consistent requests**. No notion of the primary replica -- no Scylla-side request forwarding. Drivers send request round robin to the nearest alive replicas. Scylla just checks the replicas hash - respond with the new replica list if changed.
> 2. **LWT requests**. The primary replica is chosen by the driver as in [LWT better load balancing](https://github.com/scylladb/scylla-drivers/issues/38). It seems we don't need Scylla-side request forwarding in this case.
> 3. **strongly consistent requests**. The primary replica is chosen by Scylla, put in the first position in the replica list we send in `tablets-routing-v2`, after we compute the hash. Do the Scylla-side request forwarding if the actual replica doesn't match.
